### PR TITLE
A class- / protocol-oriented design for callbacks.

### PR DESCRIPTION
**Pull request by saeta**
There are a lot of different ways to design a callback API. This change explores
one direction of having a well-typed API. There are a lot of different tradeoffs
and I'm especially interested in getting everyone's great ideas together.

If I understand correctly, the callbacks in the fast.ai python framework can
control the training loop, and even modify the model or training data. This thus
results in a variety of API design choices. One principle we adopt is that the
callbacks should be well typed, and take specific arguments and return well
defined responses. That said, there are a number of ways to actually define the
interface. To illustrate them, we pick the callback that can inspect the batch
of training data, and optionally modify it. Below are a few alternates:

 1. `func preprocessBatch(batch: Model.Input) -> Model.Input?`
   In this spelling, the callback takes the batch, and optionally returns it.
   This allows the callback to optionally modify the batch, as well as return
   `nil` to short-circuit the training loop.

2.  `func preprocessBatch(batch: Model.Input) -> BatchResult<Model.Input>`
   In this spelling, the callback takes the batch, and returns a value from the
   enum `BatchResult`, which can take values: `.continue(input: Model.Input)`,
   `.skip` which skips the batch and continues on to the next minibatch, and
   `.stop` which exits the training loop. This formulation gives a lot more
   expressiveness to the callbacks to control the training loop.

    Variant: add a `.continueUnmodified` so the callback can be clearer in the
    default case when it doesn't modify the training data.

3.  `func preprocessBatch(batch: Model.Input) -> Model.Input throws`
   In this spelling, the callback can throw in order to signal a non-default
   condition and/or control the training loop. We could define special values
   that the callback should throw.

All of the above require the callback formulations assume the batch is passed
by value. However, we could alternately use `inout`, as this simplifies the
type signatures for the return values. (e.g. No need to redundantly specify the
`Model.Input` type in the return value in addition to the parameter type.) Note
that `inout` is short for copy-in-copy-out (with optimizations), which is
exactly the semantics we want. We thus have a few alternatives for how to write
the APIs with `inout`:

4.  `func preprocessBatch(batch: inout Model.Input) -> bool`
   Return a boolean to continue or not. This has a couple disadvantages
   including not being able to halt the training loop (or skip the batch), but
   perhaps more importantly, it's confusing to the users of the API (both
   implementers as well as consumers of the API) whether `true` means continue
   or `true` means skip/halt.

5.  This final alternative is the one proposed in the PR, where the function is
    written: `func preprocessBatch(batch: inout Model.Input -> CallbackResult`
    where `CallbackResult` is an enum with the possible values of: `.continue`,
    `.stop`, and `.skip`.

That said, there are a number of open design questions that we should carefully
evaluate including (but not limited to):

 1. Would it make sense to group some of the callbacks together because they're
    often overridden together?
 2. Should this be a protocol instead of a class? (Note: we'd have to use
    associated types, which has implications on the ergonomics of usage.)
 3. How can we gracefully handle composition of multiple sets of callbacks? In
    the comments I propose chaining them together serially, which some
    reasonable way to aggregate the `CallbackResult`s.

### CODE REVIEW

**Comment by saeta** [comment]
@lattner: Would appreciate your further thoughts on the above ^^^.

**Comment by dan-zheng** [comment]
One specific comment: the existing code for `TrainingCallbacks` and `Trainer` are very rough.

Individual callbacks should be changed from properties with type `(() -> Void)` to functions:
```swift
// Change this:
open class TrainerCallback<Model : Layer, Opt : Optimizer> {
    public typealias Callback = () -> Void
    open var begin_batch: Callback? { return nil }
    open var after_pred: Callback? { return nil }
    ...
}

// To this:
open class TrainerCallback<Model : Layer, Opt : Optimizer> {
    open func beginBatch() {}
    open var afterPrediction() {}
     ...
}
```

Then, `Trainer.invokeCallback` should be updated accordingly:
```swift
// Change this:
extension Trainer {
    func invokeCallback(_ kind: CallbackKind) {
        for callback in callbacks {
            guard let cb = callback.callback(kind) else { continue }
            cb()
        }
    }
}

// To this:
extension Trainer {
    func invokeCallback(_ kind: CallbackKind) {
        for callback in callbacks {
            // Just call the callback. If it's not overridden, it will be a no-op.
            callback.callback(kind)
        }
    }
}
```

This significantly simplifies callback definitions:
```swift
final class TrainingEvaluationCallback ... {
    // Change this:
    override var begin_fit: Callback {
        return {
            self.epochCount = 0
            self.iterationCount = 0
        }
    } 

    // To this:
    override func beginFit {
        self.epochCount = 0
        self.iterationCount = 0
    } 
}
```

---

Does this design make sense to everyone?

**Comment by dan-zheng** [comment]
Another aside, sorry: a bigger issue mentioned by @rxwei is that trainer abstractions represented as "structs storing a value-typed model" are problematic.

During initialization, such trainers would store a copy of the original input model, so the original model would never mutated. Storing `inout Model` is not possible.

Future design is necessary. So far, we've represented training loops as functions taking an `inout Model` argument.

**Comment by dan-zheng** [comment]
> If I understand correctly, the callbacks in the fast.ai python framework can
> control the training loop, and even modify the model or training data.

I think the main guiding question for the callback design is: are there in fact example callbacks that mutate data and/or models, or do callback typically not mutate things?

cc @jph00 @sgugger: any examples and insight would be a big help!

**Comment by sgugger** [comment]
I don't have much time since there is a course in three hours, but the short answer is yes. You can't have a system of Callbacks that allow any modification in the training loop if you restrain the Callbacks. They must have access (and be able to change) anything including the model, the loss function, the optimizer (the general context we put in a Learner object) and the last input/target/output/loss/gradients.

All the modules under callbacks in the fastai library can illustrate that, but to give a few specific examples:
- mixup needs to change the loss function and replace it (it also needs to change the inputs/targets on the fly at each batch)
- the callbacks that schedule hyperparameters need access to the optimizer to change those values in its state
- a callback might change the whole dataset when you do progressive resizing (so you change the sizes of your images from 128 to 256 for instance) or change the data augmentation applied to the inputs
- a callback that trains GANs need access to the model to be able to switch between generator and critic
- not sure if mixed precision training requires to change the model to put the weights in half precision and the whole optimizer as it changes the parameters that are optimized

And that's just what is on the top of my head, to be sure to support any tweak researchers come with we really need to have access to anything in a Callback and be allowed to modify it. 
As a side note, we found it's easier to have all that information in one (mutable) object passed to the Callback (the Runner in 04, that becomes the Learner in 09b).

**Comment by saeta** [comment]
Thanks @sgugger that's really helpful!

One related question is: what are all the callbacks that make sense? For now, I can see (better names wanted; see below):
 - `beforeStart`
 - `beforeEpoch` (does this even make sense?)
 - `beforeBatch`
 - `preprocessBatch` (Note: it should take at least: `(inputs: inout Model.Input, labels: inout Model.Output)`, which I definitely missed).
 - `preprocessPrediction`
 - `preprocessLoss`
 - `preprocessGradients`
 - `afterBatch`
 - `afterEpoch`
 - `afterFinished`

Related note: what's our policy on naming these callbacks? Do we want to name them for when they're called? Or for what they [can] do? (One advantage for the latter is that users might not need to know quite so much about implementation details of the training loop. That said, these names will definitely require a lot of care // might not make sense.)

Walking through your examples of how to implement them:
 1. *mixup*: By overriding `preprocessBatch`, it could perform the mixing of the inputs and labels. (Note: we need to think carefully about how to handle the one-hot encoding; mixup would require the data type to not be a 1-D tensor of `Int`s anymore.) One open question is how to handle the loss [function changes](https://github.com/fastai/fastai/blob/master/fastai/callbacks/mixup.py#L39).

 2. *hyperparameter schedules*: The callbacks should definitely be able to get a pointer to the optimizer (either because the user passed it in at callback construction time, or because we pass it to the callbacks in some form) and [set hyperparameters](https://github.com/tensorflow/swift-apis/blob/master/Sources/DeepLearning/Optimizer.swift#L48). So this is pretty reasonable. One thing to note: we might need to change the `Optimizer` protocol (as it currently only allows [retrieving the learning rate](https://github.com/tensorflow/swift-apis/blob/master/Sources/DeepLearning/Optimizer.swift#L29)) or alternately, define a protocol refinement to `Optimizer` (such as `ScheduleableOptimizer`) which would allow setting the learning rate or other common hyperparameters. Alternately, we could have an implementation of the learning rate scheduler specialized for each specific optimizer.

 3. *progressive resizing / data augmentation*: This looks doable with the suggested callback API. Concretely, assuming the input pipeline is changed to output the largest size images, the callback could do the scaling down. (Further design and discussion required to avoid the performance (and potential quality) hit of a double-resize. But this is more in the design of the data loader API, which I think we can make reasonably flexible. In fact, does it make most sense to support this purely from the data loading side, instead of via callbacks?)

 4. *GAN callback to switch between generator & critic*: Hmmm, I'm not quite sure how to think about this one, as presumably these are different models (and thus different types). We could type erase them by hiding things behind a protocol (although I think this would interact poorly with the autodiff system). Further thought required here (also see the next comment).

 5. *mixed precision / device scheduling*: For this one, I think changing the model precision makes a lot of sense to be done via some other mechanism than a callback. As we think about how to do efficient code-generation targeting a variety of different accelerators and platforms (including mobile), we probably want to have this as part of the type. (Specifically, we would have `ResNet<Float>` vs `ResNet<BFloat16>` vs `ResNet<Int8>` for the different formats. Thinking aloud, we might need two type parameters, one for the model weights, and one for the activations... although we'll need to carefully think through the quantization story.)

Taking a step back, I think the more use cases we can validate our design with, the better so having more applications like the above are super appreciated. I also want to take a step back and try and distill the flexibility needed so we're thinking not just through specific cases but try and make generalizations about the kinds of things you can and can't do with the different API formulations.

I think we should also plan on further iteration and refinement of our designs over time. :-)

All the best,
-Brennan

**Comment by lattner** [comment]
saeta@ your description above is great, thank you for walking through this!   @dan-zheng we know that the current design is suboptimal in a couple of ways.  Brennan is walking through one set of tradeoffs that are orthogonal from the others.

**Comment by jph00** [comment]
> One related question is: what are all the callbacks that make sense? 

The best way to answer that question is to look at notebook 09b where you'll see them in context. Here's a snapshot (all the bits that look like `self("something")` are calls to callbacks):

```bash
$ ack "self\('" nb_09b.py -C1
            self.xb,self.yb = xb,yb
            self('begin_batch')
            self.pred = self.model(self.xb)
            self('after_pred')
            self.loss = self.loss_func(self.pred, self.yb)
            self('after_loss')
            if not self.in_train: return
            self.loss.backward()
            self('after_backward')
            self.opt.step()
            self('after_step')
            self.opt.zero_grad()
        except CancelBatchException: self('after_cancel_batch')
        finally: self('after_batch')

--
            for xb,yb in dl: self.one_batch(xb, yb)
        except CancelEpochException: self('after_cancel_epoch')

--
            for cb in self.cbs: cb.set_runner(self)
            self('begin_fit')
            for epoch in range(epochs):
                self.epoch = epoch
                if not self('begin_epoch'): self.all_batches(self.data.train_dl)

                with torch.no_grad():
                    if not self('begin_validate'): self.all_batches(self.data.valid_dl)
                self('after_epoch')

        except CancelTrainException: self('after_cancel_train')
        finally: self('after_fit')
```

As you'll see from the context in the notebook, there's a callback in between every step of the training loop, and before and after every section. The reasoning is already explained in detail in the lessons so I won't try to repeat it here.

> preprocessBatch (Note: it should take at least: (inputs: inout Model.Input, labels: inout Model.Output), which I definitely missed)

As @sgugger mentioned, the callbacks need to be able to access and change everything. If you try to pass and return just some subset of data, loss func, optimizer, model, and batch to each callback based on how you guess people will use it, you'll find people keep finding ways to break your assumptions. We know this from bitter experience! ;) Maybe the right answer isn't callbacks at all - but somehow, users need to be able to inject behavior into the training loop which can inspect and modify all the things mentioned above at any time.

> beforeEpoch (does this even make sense?)

Yes it does (e.g. to only do validation and not training you return `True` from it, or to calibrate per-epoch stuff before you start an epoch you need it, etc). But the more important reasoning is that you need to provide injection points in between every step of the training loop, and before and after every section, otherwise you're second-guessing what you think people need, and you end up getting unpleasantly surprised by your users that you missed something (as we've discovered ourselves!)

> Related note: what's our policy on naming these callbacks? Do we want to name them for when they're called? Or for what they [can] do?

They don't do anything, and they can do everything, so they can't be named by either of those things. So they should be named by when they're called. The ones you've called `preprocess*` can (and are) used for things other than preprocessing those things, for instance, 

> One advantage for the latter is that users might not need to know quite so much about implementation details of the training loop.

You can't really extend a training loop without understanding what it does at each point. See our docs for callback writers to see how we communicate this in practice: https://docs.fast.ai/callback.html#Methods-your-subclass-can-implement

> So this is pretty reasonable. One thing to note: we might need to change the Optimizer protocol (as it currently only allows retrieving the learning rate) or alternately, define a protocol refinement to Optimizer (such as ScheduleableOptimizer) which would allow setting the learning rate or other common hyperparameters.

The minimal optimizer capabilities you'll need are those shown in notebook 09 `StatefulOptimizer`. That notebook also shows how these pieces are used to create various other specific optimizers, including the *LAMB* optimizer that Google published a few days ago (in the "BERT in 76 minutes" paper).

> For this one, I think changing the model precision makes a lot of sense to be done via some other mechanism than a callback.

Mixed precision training is different to the issues around bfloat16/int8 training. It's a specific set of steps shown in notebook 12. It's very hard to do it without callbacks. Nvidia have tried, and have recently released "Apex" which attempts to automate this - but because they didn't use callbacks they've ended up with something that's fragile and tightly coupled to one optimizer and one training loop. It also takes them a *lot* of code.

> I think the more use cases we can validate our design with, the better so having more applications like the above are super appreciated.

Here's a list of all of our callbacks: https://docs.fast.ai/callbacks.html . (There's also one other callback that's mentioned in the `basic_train` docs, which is [Recorder](https://docs.fast.ai/basic_train.html#Recorder).)

Having said that, it might be easier to just focus on getting the notebooks working for now, which means you need to support these (from the course repo):

```
$ ack 'class.*Callback'
nb_05.py
42:class Recorder(Callback):
55:class ParamScheduler(Callback):

nb_05b.py
9:class Callback():
24:class TrainEvalCallback(Callback):
113:class AvgStatsCallback(Callback):
129:class Recorder(Callback):
142:class ParamScheduler(Callback):

nb_12.py
9:class MixedPrecision(Callback):

nb_06.py
23:class CudaCallback(Callback):
27:class BatchTransformXCallback(Callback):
```

**Comment by jph00** [comment]
@saeta: Can we close this now since the Learner from @marcrasi seems to largely obsolete these changes?

**Comment by saeta** [comment]
Yup. Closing this now.
